core/server/master/src/main/java/alluxio/master/metrics/DefaultMetricsMaster.java

```diff
@@ -188,10 +193,17 @@ public String getName() {
   public void start(Boolean isLeader) throws IOException {
     super.start(isLeader);
     if (isLeader) {
+      mMetricsStore.start();
       getExecutorService().submit(mClusterMetricsUpdater);
     }
   }
 
+  @Override
+  public void stop() throws IOException {
+    super.stop();
+    mMetricsStore.stop();
```

Comment:
I have a high-level question. Should the queue and async nature be in the MetricsMaster, instead of the MetricsStore? It might make sense to make the MetricsStore simply a storing data structure, and this MetricsMaster just happens to update the store in an async way? I feel that allows the MetricsStore to simply handle storing/serving the metrics data as efficiently as possible.

What are you thoughts on that?

Reply:
Do you mean that we MetricsMaster have 
1) A queue storing all the worker/client metric submit through heartbeat
2) A thread pool (executor), if detect any element in the queue, poll from queue and submit to the executor which calls metricsStore.processMetrics

In this way, metricsStore is a storing data structure and it's multi-threaded instead of single-threaded. (More efficiency if we improve the Metric concurrency)

Reply:
I think that is what I was proposing.

Yes, I was proposing that the MetricsMaster owns the threads, and executors, to implement the async nature of the updates.

I'm not sure what you mean by `metricsStore is a storing data structure and it's multi-threaded`. I was saying that the MetricsStore is a data structure, and the usage of that data structure (MetricsMaster) will call it in an async or multi-threaded way.